vllm/v1/worker/gpu_model_runner.py

```diff
@@ -1643,6 +1703,35 @@ def _preprocess(
                 **self._init_model_kwargs(num_scheduled_tokens),
                 **self._extract_mm_kwargs(scheduler_output),
             }
+        elif (self.enable_prompt_embeds and get_pp_group().is_first_rank):
+            # Get the input embeddings for the tokens that are not input embeds,
+            # then put them into the appropriate positions.
+            # TODO(qthequartermasterman): Since even when prompt embeds are
+            # enabled, (a) not all requests will use prompt embeds, and (b)
+            # after the initial prompt is processed, the rest of the generated
+            # tokens will be token ids, it is not desirable to have the
+            # embedding layer outside of the CUDA graph all the time. The v0
+            # engine avoids this by "double compiling" the CUDA graph, once
+            # with input_ids and again with inputs_embeds, for all num_tokens.
+            # If a batch only has token ids, then including the embedding layer
+            # in the CUDA graph will be more performant (like in the else case
+            # below).
+            token_ids_idx = self.is_token_ids.gpu[:num_scheduled_tokens] \
+                .nonzero(as_tuple=False) \
+                .squeeze(1)
+            # Some tokens ids may need to become embeds
+            # We use index_select/index_copy_ instead of slicing to make sure
+            # the copy happens in place.
+            if token_ids_idx.numel() > 0:
+                token_ids = self.input_ids.gpu.index_select(0, token_ids_idx)
+                tokens_to_embeds = self.model.get_input_embeddings(
+                    input_ids=token_ids)
+                self.inputs_embeds.gpu.index_copy_(0, token_ids_idx,
+                                                   tokens_to_embeds)
```

Comment:
```suggestion
                token_ids = self.input_ids.gpu[token_ids_idx]
                tokens_to_embeds = self.model.get_input_embeddings(
                    input_ids=token_ids)
                self.inputs_embeds.gpu[token_ids_idx] = tokens_to_embeds
```

QQ: How is your code different from this?

Reply:
That is quite surprising, I thought slice assignment always applied to the original tensor.

Reply:
After pondering this more, you are actually right. Looking at what I was doing before wasn't quite the same as what you suggest. I was using `copy_` instead of an advanced index assignment. The advanced index assignment does indeed function in place, whereas I was doing `copy_` on what was sometimes a view and was sometimes a copy depending on the contents of token_ids_idx.

I accepted your change. One difference could be how casting is done. `index_copy_` handles casting to the correct device/dtype when needed, but it looks like that shouldn't be a problem in this context because tokens_to_embeds should already be on the right device/dtype.